pkg/registry/scheduling/rest/storage_scheduling.go

```diff
@@ -89,51 +95,99 @@ func (p RESTStorageProvider) v1Storage(apiResourceConfigSource serverstorage.API
+		}
+		// Add default resource quota
+		if err := AddDefaultSystemQuota(hookContext); err != nil {
+			return fmt.Errorf("unable to add default system resource quota: %v", err)
+		}
+		return nil
+	}
+}
+
+func AddSystemPriorityClasses(hookContext genericapiserver.PostStartHookContext) error {
+	return wait.Poll(1*time.Second, 30*time.Second, func() (done bool, err error) {
+		schedClientSet, err := schedulingclient.NewForConfig(hookContext.LoopbackClientConfig)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("unable to initialize client: %v", err))
+			return false, nil
+		}
 
-			for _, pc := range scheduling.SystemPriorityClasses() {
-				_, err := schedClientSet.PriorityClasses().Get(pc.Name, metav1.GetOptions{})
-				if err != nil {
-					if apierrors.IsNotFound(err) {
-						// TODO: Remove this explicit conversion after scheduling api move to v1
-						v1beta1PriorityClass := &schedulingv1beta1.PriorityClass{}
-						if err := schedulingapiv1beta1.Convert_scheduling_PriorityClass_To_v1beta1_PriorityClass(pc, v1beta1PriorityClass, nil); err != nil {
-							return false, err
-						}
-						_, err := schedClientSet.PriorityClasses().Create(v1beta1PriorityClass)
-						if err != nil && !apierrors.IsAlreadyExists(err) {
-							return false, err
-						} else {
-							klog.Infof("created PriorityClass %s with value %v", pc.Name, pc.Value)
-						}
+		for _, pc := range scheduling.SystemPriorityClasses() {
+			_, err := schedClientSet.PriorityClasses().Get(pc.Name, metav1.GetOptions{})
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					// TODO: Remove this explicit conversion after scheduling api move to v1
+					v1beta1PriorityClass := &schedulingv1beta1.PriorityClass{}
+					if err := schedulingapiv1beta1.Convert_scheduling_PriorityClass_To_v1beta1_PriorityClass(pc, v1beta1PriorityClass, nil); err != nil {
+						return false, err
+					}
+					_, err := schedClientSet.PriorityClasses().Create(v1beta1PriorityClass)
+					if err != nil && !apierrors.IsAlreadyExists(err) {
+						return false, err
 					} else {
-						// Unable to get the priority class for reasons other than "not found".
-						klog.Warningf("unable to get PriorityClass %v: %v. Retrying...", pc.Name, err)
-						return false, nil
+						klog.Infof("created PriorityClass %s with value %v", pc.Name, pc.Value)
 					}
+				} else {
+					// Unable to get the priority class for reasons other than "not found".
+					klog.Warningf("unable to get PriorityClass %v: %v. Retrying...", pc.Name, err)
+					return false, nil
 				}
 			}
-			klog.Infof("all system priority classes are created successfully or already exist.")
-			return true, nil
-		})
-		// if we're never able to make it through initialization, kill the API server.
+		}
+		klog.Infof("all system priority classes are created successfully or already exist.")
+		return true, nil
+	})
+}
+
+func AddDefaultSystemQuota(hookContext genericapiserver.PostStartHookContext) error {
```

Comment:
This seems like a significant change.  Can you document why we are going to have an undeleteable resourcequota object?  Is this trying to preserve some previous behavior around allowing any number of critical pods in the kube-system namespace?  I'm not seeing a value for the number of pods set, so it's not obvious to me what it's doing.

Reply:
>  Is this trying to preserve some previous behavior around allowing any number of critical pods in the kube-system namespace? I'm not seeing a value for the number of pods set, so it's not obvious to me what it's doing.

Yeah, it's to ensure backwards compatibility to allow unlimited critical pods in `kube-system` namespace. This is especially true for DS critical DS pods, which as of now are being created in kube-system namespace.

>  Can you document why we are going to have an undeleteable resourcequota object?

I will create a document.